### PR TITLE
Voxel51 Integration

### PR DESCRIPTION
**Pull request by aaron-siegel**

### CODE REVIEW

**Comment by aaron-siegel** [comment]
This change is [Reviewable](https://reviewable.io/reviews/pixeltable/pixeltable/337)

File: pixeltable/dataframe.py [hunk 1]
```diff
@@ -371,15 +371,10 @@ def bind(self, args: dict[str, Any]) -> DataFrame:
             group_by_clause=group_by_clause, grouping_tbl=self.grouping_tbl,
             order_by_clause=order_by_clause, limit=self.limit_val)
 
-    def collect(self) -> DataFrameResultSet:
-        return self._collect()
-
-    def _collect(self, conn: Optional[sql.engine.Connection] = None) -> DataFrameResultSet:
+    def _row_iterator(self, conn: Optional[sql.engine.Connection] = None) -> Iterator[list]:
```
Comment (mkornacker) [inline pixeltable/dataframe.py]: There are a lot of references to 'rows' everywhere. Should probably call this _output_row_iterator for clarity.

File: pixeltable/io/fiftyone.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+from pixeltable import exprs
+from pixeltable.env import Env
+
+
+class PxtImageDatasetImporter(foud.LabeledImageDatasetImporter):
+    """
+    Implementation of a FiftyOne `DatasetImporter` that reads image data from a Pixeltable table.
+    """
+    __image_format: str
+    __labels: dict[str, tuple[exprs.Expr, type]]  # label_name -> (expr, label_cls)
```
Comment (mkornacker) [inline pixeltable/io/fiftyone.py]: Is there a base class for label_cls?

File: pixeltable/io/fiftyone.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+    __image_idx: int
```
Comment (mkornacker) [inline pixeltable/io/fiftyone.py]: idx into what?

File: pixeltable/io/fiftyone.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+    __localpath_idx: Optional[int]
```
Comment (mkornacker) [inline pixeltable/io/fiftyone.py]: same here

File: pixeltable/io/fiftyone.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+    __count: int
```
Comment (mkornacker) [inline pixeltable/io/fiftyone.py]: of what?

File: pixeltable/io/fiftyone.py [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+    __iter: Iterator[list]
```
Comment (mkornacker) [inline pixeltable/io/fiftyone.py]: over what?

File: pixeltable/io/globals.py [hunk 7]
```diff
@@ -267,3 +273,90 @@ def import_json(
         contents = urllib.request.urlopen(filepath_or_url).read()
     data = json.loads(contents, **kwargs)
     return import_rows(tbl_path, data, schema_overrides=schema_overrides, primary_key=primary_key, num_retained_versions=num_retained_versions, comment=comment)
+
+
+def export_images_as_fiftyone(
```
Comment (mkornacker) [inline pixeltable/io/globals.py]: _to_fiftyone()? create_fiftyone_labeledimagedataset() to be completely clear?

File: pixeltable/io/globals.py [hunk 8]
```diff
@@ -267,3 +273,90 @@ def import_json(
...
+    tbl: pxt.Table,
+    images: exprs.Expr,
+    image_format: str = 'webp',
+    classifications: Union[exprs.Expr, list[exprs.Expr], dict[str, exprs.Expr], None] = None,
+    detections: Union[exprs.Expr, list[exprs.Expr], dict[str, exprs.Expr], None] = None,
+) -> 'fo.Dataset':
```
Comment (mkornacker) [inline pixeltable/io/globals.py:284]: -> LabeledImageDataset?

Probably good to tell the user exactly what they're getting, given that 51 has this intricate class hierarchy.

**Review by mkornacker — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pixeltable/pixeltable/337)* status: 0 of 13 files reviewed, 11 unresolved discussions (waiting on @aaron-siegel)

___
*[`pixeltable/io/fiftyone.py` line 97 at r3](https://reviewable.io/reviews/pixeltable/pixeltable/337#-OA1l4rT5RCiOkTkP2DS:-OA1l4rT5RCiOkTkP2DT:b6d3en8) ([raw file](https://github.com/pixeltable/pixeltable/blob/c6dbc52acd2691d146e20736db8869df35a108e8/pixeltable/io/fiftyone.py#L97)):*
> ```Python
> 
>         df = tbl.select(*selection)
>         self.__count = df.count()
> ```

Is this necessary? This could double the runtime of this operation.
___
*[`pixeltable/io/fiftyone.py` line 146 at r3](https://reviewable.io/reviews/pixeltable/pixeltable/337#-OA1mRe7BG6iZacBIqUV:-OA1mRe7BG6iZacBIqUW:b-3z3txo) ([raw file](https://github.com/pixeltable/pixeltable/blob/c6dbc52acd2691d146e20736db8869df35a108e8/pixeltable/io/fiftyone.py#L146)):*
> ```Python
>             raise excs.Error(f'Invalid classifications data: {data}')
>         return [
>             fo.Classification(label=entry['label'], confidence=entry.get('confidence'))
> ```

What if there's no label field?
___
*[`pixeltable/io/fiftyone.py` line 154 at r3](https://reviewable.io/reviews/pixeltable/pixeltable/337#-OA1mWPb7DRQLWnFoJGp:-OA1mWPb7DRQLWnFoJGq:b-5k5bq6) ([raw file](https://github.com/pixeltable/pixeltable/blob/c6dbc52acd2691d146e20736db8869df35a108e8/pixeltable/io/fiftyone.py#L154)):*
> ```Python
>             raise excs.Error(f'Invalid detections data: {data}')
>         return [
>             fo.Detection(label=entry['label'], bounding_box=entry['bounding_box'], confidence=entry.get('confidence'))
> ```

same here for bounding_box

File: pixeltable/io/globals.py [hunk 9]
```diff
@@ -267,3 +273,90 @@ def import_json(
...
+
+        See the [Working with Voxel51 in Pixeltable](https://docs.pixeltable.com/docs/working-with-voxel51) tutorial
+        for a fully worked example.
+    """
+    Env.get().require_package('fiftyone')
+
+    import fiftyone as fo
+    from pixeltable.io.fiftyone import PxtImageDatasetImporter
+
+    if not images.col_type.is_image_type():
```
Comment (mkornacker) [inline pixeltable/io/globals.py:358]: Additional checks:
- img expr should be bound by table (see Expr.is_bound_by())
- same for classification and detection exprs

**Review by aaron-siegel — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pixeltable/pixeltable/337)* status: 0 of 13 files reviewed, 12 unresolved discussions (waiting on @mkornacker)

___
*[`pixeltable/dataframe.py` line 374 at r2](https://reviewable.io/reviews/pixeltable/pixeltable/337#gh-1815164657:-OA3l0MjCQ_0MuOB5q1K:b-896fix) ([raw file](https://github.com/pixeltable/pixeltable/blob/3ae49c1362d3a531b09154eb510c44d50433aef1/pixeltable/dataframe.py#L374)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

There are a lot of references to 'rows' everywhere. Should probably call this _output_row_iterator for clarity.
</blockquote></details>

Done.
___
*[`pixeltable/io/fiftyone.py` line 20 at r2](https://reviewable.io/reviews/pixeltable/pixeltable/337#gh-1815175540:-OA3lQFn0II3Pvcccfxz:brssq08) ([raw file](https://github.com/pixeltable/pixeltable/blob/3ae49c1362d3a531b09154eb510c44d50433aef1/pixeltable/io/fiftyone.py#L20)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

Is there a base class for label_cls?
</blockquote></details>

Yes, `fo.Label`. Updated to make this more specific.
___
*[`pixeltable/io/fiftyone.py` line 21 at r2](https://reviewable.io/reviews/pixeltable/pixeltable/337#gh-1815176257:-OA3lerD-1pFHu7RQy4R:b-u5t80l) ([raw file](https://github.com/pixeltable/pixeltable/blob/3ae49c1362d3a531b09154eb510c44d50433aef1/pixeltable/io/fiftyone.py#L21)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

idx into what?
</blockquote></details>

Added comment to clarify.
___
*[`pixeltable/io/fiftyone.py` line 22 at r2](https://reviewable.io/reviews/pixeltable/pixeltable/337#gh-1815177493:-OA3mBmVAqNc48EzOINb:b-bizqh8) ([raw file](https://github.com/pixeltable/pixeltable/blob/3ae49c1362d3a531b09154eb510c44d50433aef1/pixeltable/io/fiftyone.py#L22)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

same here
</blockquote></details>

Added comment.
___
*[`pixeltable/io/fiftyone.py` line 23 at r2](https://reviewable.io/reviews/pixeltable/pixeltable/337#gh-1815178151:-OA3mCiOAvjwCmz-SqU1:b-rv4jec) ([raw file](https://github.com/pixeltable/pixeltable/blob/3ae49c1362d3a531b09154eb510c44d50433aef1/pixeltable/io/fiftyone.py#L23)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

of what?
</blockquote></details>

Changed to `__dataset_size` and added comment.
___
*[`pixeltable/io/fiftyone.py` line 24 at r2](https://reviewable.io/reviews/pixeltable/pixeltable/337#gh-1815178474:-OA3mU7y02OnQ1AI9ng7:bf8fgox) ([raw file](https://github.com/pixeltable/pixeltable/blob/3ae49c1362d3a531b09154eb510c44d50433aef1/pixeltable/io/fiftyone.py#L24)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

over what?
</blockquote></details>

Changed to `__row_iter` and added comment.
___
*[`pixeltable/io/fiftyone.py` line 97 at r3](https://reviewable.io/reviews/pixeltable/pixeltable/337#-OA1l4rT5RCiOkTkP2DS:-OA3mXyI4FqoHQbwkuiQ:b-7bpawx) ([raw file](https://github.com/pixeltable/pixeltable/blob/c6dbc52acd2691d146e20736db8869df35a108e8/pixeltable/io/fiftyone.py#L97)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

Is this necessary? This could double the runtime of this operation.
</blockquote></details>

Yeah, I understand. The docs seem to imply that Voxel51 expects `__len__()` to be implemented, but I just tried removing it and the notebook and tests still work. So maybe we leave it out for now, and hope for the best?
___
*[`pixeltable/io/fiftyone.py` line 146 at r3](https://reviewable.io/reviews/pixeltable/pixeltable/337#-OA1mRe7BG6iZacBIqUV:-OA3nmWEEqCE-RhEWQ7P:b-a0gn9u) ([raw file](https://github.com/pixeltable/pixeltable/blob/c6dbc52acd2691d146e20736db8869df35a108e8/pixeltable/io/fiftyone.py#L146)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

What if there's no label field?
</blockquote></details>

It will throw an exception. I added a clause to generate an `excs.Error` rather than a `KeyError`, with a clearer error message
___
*[`pixeltable/io/fiftyone.py` line 154 at r3](https://reviewable.io/reviews/pixeltable/pixeltable/337#-OA1mWPb7DRQLWnFoJGp:-OA3oaaK35FhcA4RzWBH:bb73xfv) ([raw file](https://github.com/pixeltable/pixeltable/blob/c6dbc52acd2691d146e20736db8869df35a108e8/pixeltable/io/fiftyone.py#L154)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

same here for bounding_box
</blockquote></details>

Same
___
*[`pixeltable/io/globals.py` line 278 at r3](https://reviewable.io/reviews/pixeltable/pixeltable/337#gh-1816204454:-OA3pA6n0ZN7iPF9Wmh_:bxd98uz) ([raw file](https://github.com/pixeltable/pixeltable/blob/c6dbc52acd2691d146e20736db8869df35a108e8/pixeltable/io/globals.py#L278)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

_to_fiftyone()? create_fiftyone_labeledimagedataset() to be completely clear?
</blockquote></details>

I changed it to `export_images_to_fiftyone_dataset`. I don't have a strong preference between `to` and `as`, but it seems like `as` is what we've used more typically for conversions? (`as_dict`, `astype`, etc)

`create_fiftyone_labeledimagedataset` is a big mouthful, and I'd argue `export` better captures what's happening than `create`
___
*[`pixeltable/io/globals.py` line 284 at r3](https://reviewable.io/reviews/pixeltable/pixeltable/337#gh-1816204564:-OA3pi3MC7o7qIRYrkAS:bdhn9ii) ([raw file](https://github.com/pixeltable/pixeltable/blob/c6dbc52acd2691d146e20736db8869df35a108e8/pixeltable/io/globals.py#L284)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

-> LabeledImageDataset?

Probably good to tell the user exactly what they're getting, given that 51 has this intricate class hierarchy.
</blockquote></details>

there is no such thing as a "LabeledImageDataset" afaik. (the `DatasetImporter`s have a big class hierarchy, but there's only one kind of `Dataset`)

```
In [11]: type(fo_dataset)
Out[11]: fiftyone.core.dataset.Dataset
```
___
*[`pixeltable/io/globals.py` line 357 at r3](https://reviewable.io/reviews/pixeltable/pixeltable/337#gh-1816243662:-OA3qowmDNNUwtbxkJL2:b-fdspdy) ([raw file](https://github.com/pixeltable/pixeltable/blob/c6dbc52acd2691d146e20736db8869df35a108e8/pixeltable/io/globals.py#L357)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

Additional checks:
- img expr should be bound by table (see Expr.is_bound_by())
- same for classification and detection exprs
</blockquote></details>

Do we really want to do those checks here?

I'd argue that we'd be better off enforcing those checks in `t.select()` (which has been a backlog item for awhile now) and letting the nested call to `select()` do the error handling.

**Review by mkornacker — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pixeltable/pixeltable/337)* status: 0 of 13 files reviewed, 12 unresolved discussions (waiting on @aaron-siegel)

___
*[`pixeltable/io/globals.py` line 278 at r3](https://reviewable.io/reviews/pixeltable/pixeltable/337#gh-1816204454:-OA_LDzWBlpQgbk0COF3:b-tpgozi) ([raw file](https://github.com/pixeltable/pixeltable/blob/c6dbc52acd2691d146e20736db8869df35a108e8/pixeltable/io/globals.py#L278)):*
<details><summary><i>Previously, aaron-siegel (Aaron Siegel) wrote…</i></summary><blockquote>

I changed it to `export_images_to_fiftyone_dataset`. I don't have a strong preference between `to` and `as`, but it seems like `as` is what we've used more typically for conversions? (`as_dict`, `astype`, etc)

`create_fiftyone_labeledimagedataset` is a big mouthful, and I'd argue `export` better captures what's happening than `create`
</blockquote></details>

So then maybe just export_as_fo_dataset()?
___
*[`pixeltable/io/globals.py` line 284 at r3](https://reviewable.io/reviews/pixeltable/pixeltable/337#gh-1816204564:-OA_KsWS0Befz1ESLA5P:b-jpr3hm) ([raw file](https://github.com/pixeltable/pixeltable/blob/c6dbc52acd2691d146e20736db8869df35a108e8/pixeltable/io/globals.py#L284)):*
<details><summary><i>Previously, aaron-siegel (Aaron Siegel) wrote…</i></summary><blockquote>

there is no such thing as a "LabeledImageDataset" afaik. (the `DatasetImporter`s have a big class hierarchy, but there's only one kind of `Dataset`)

```
In [11]: type(fo_dataset)
Out[11]: fiftyone.core.dataset.Dataset
```
</blockquote></details>

Hm, weird. I found this
https://docs.voxel51.com/api/fiftyone.types.dataset_types.html#fiftyone.types.dataset_types.LabeledImageDataset

but I don't know how that relates to Dataset
___
*[`pixeltable/io/globals.py` line 357 at r3](https://reviewable.io/reviews/pixeltable/pixeltable/337#gh-1816243662:-OA_L_D0A-_vfoEwqFW9:bg48mer) ([raw file](https://github.com/pixeltable/pixeltable/blob/c6dbc52acd2691d146e20736db8869df35a108e8/pixeltable/io/globals.py#L357)):*
<details><summary><i>Previously, aaron-siegel (Aaron Siegel) wrote…</i></summary><blockquote>

Do we really want to do those checks here?

I'd argue that we'd be better off enforcing those checks in `t.select()` (which has been a backlog item for awhile now) and letting the nested call to `select()` do the error handling.
</blockquote></details>

Do we have a jira for that? I think we should do that soon.

**Review by mkornacker — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pixeltable/pixeltable/337)* status: 0 of 13 files reviewed, 12 unresolved discussions (waiting on @aaron-siegel)

___
*[`pixeltable/io/fiftyone.py` line 97 at r3](https://reviewable.io/reviews/pixeltable/pixeltable/337#-OA1l4rT5RCiOkTkP2DS:-OA_MoVi3lXGjpkg_hUd:b6hgo7) ([raw file](https://github.com/pixeltable/pixeltable/blob/c6dbc52acd2691d146e20736db8869df35a108e8/pixeltable/io/fiftyone.py#L97)):*
<details><summary><i>Previously, aaron-siegel (Aaron Siegel) wrote…</i></summary><blockquote>

Yeah, I understand. The docs seem to imply that Voxel51 expects `__len__()` to be implemented, but I just tried removing it and the notebook and tests still work. So maybe we leave it out for now, and hope for the best?
</blockquote></details>

Let's go with that.

**Review by mkornacker — APPROVED** [review]

**Review by aaron-siegel — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pixeltable/pixeltable/337)* status: 0 of 13 files reviewed, 12 unresolved discussions (waiting on @mkornacker)

___
*[`pixeltable/io/globals.py` line 278 at r3](https://reviewable.io/reviews/pixeltable/pixeltable/337#gh-1816204454:-OAd46M63qEMpbFcV3U4:bmilplg) ([raw file](https://github.com/pixeltable/pixeltable/blob/c6dbc52acd2691d146e20736db8869df35a108e8/pixeltable/io/globals.py#L278)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

So then maybe just export_as_fo_dataset()?
</blockquote></details>

`export_images_as_fo_dataset`? We really need to specify that they're images since the video export function will have a different signature (due to different label types), and there's no way to statically distinguish a video expression from an image expression
___
*[`pixeltable/io/globals.py` line 284 at r3](https://reviewable.io/reviews/pixeltable/pixeltable/337#gh-1816204564:-OAd4UxKCJXAwcoEs2pS:bq31v9a) ([raw file](https://github.com/pixeltable/pixeltable/blob/c6dbc52acd2691d146e20736db8869df35a108e8/pixeltable/io/globals.py#L284)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

Hm, weird. I found this
https://docs.voxel51.com/api/fiftyone.types.dataset_types.html#fiftyone.types.dataset_types.LabeledImageDataset

but I don't know how that relates to Dataset
</blockquote></details>

Yeah, I think those are just type indicators for your dataset. so an `fo.Dataset` is a `fiftyone.core.dataset.Dataset`, which is not the same thing as a `fiftyone.types.dataset_types.Dataset`, and it's the latter thing that's the base class of `LabeledImageDataset`. Confusing.

**Review by aaron-siegel — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pixeltable/pixeltable/337)* status: 0 of 13 files reviewed, 12 unresolved discussions (waiting on @mkornacker)

___
*[`pixeltable/io/fiftyone.py` line 97 at r3](https://reviewable.io/reviews/pixeltable/pixeltable/337#-OA1l4rT5RCiOkTkP2DS:-OAe4XXu96DO1rkL4OGE:b-896fix) ([raw file](https://github.com/pixeltable/pixeltable/blob/c6dbc52acd2691d146e20736db8869df35a108e8/pixeltable/io/fiftyone.py#L97)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

Let's go with that.
</blockquote></details>

Done.
___
*[`pixeltable/io/globals.py` line 357 at r3](https://reviewable.io/reviews/pixeltable/pixeltable/337#gh-1816243662:-OAe4nQhDVsvgoJHev9V:b-fn2s2r) ([raw file](https://github.com/pixeltable/pixeltable/blob/c6dbc52acd2691d146e20736db8869df35a108e8/pixeltable/io/globals.py#L357)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

Do we have a jira for that? I think we should do that soon.
</blockquote></details>

Just created PXT-328.

**Pull request merged**